test/e2e/framework/conformance/conformance.go

```diff
@@ -0,0 +1,845 @@
+					return fmt.Errorf("get existing %s: %w", resourceType, err)
+				}
+			}
+			object := tc.GetUpdateSpec(existingObject.DeepCopy())
+			ginkgo.By(fmt.Sprintf("Updating:\n%s", marshalYAML(object, 1)))
+			existingObject, err = client.Update(ctx, object, metav1.UpdateOptions{})
+			if err == nil {
+				return nil
+			}
+			mustGet = apierrors.IsConflict(err)
+			if mustGet {
+				// Retry immediately.
+				return fmt.Errorf("update %s: %w", resourceType, err)
+			}
+			if retry, retryAfter := framework.ShouldRetry(err); retry {
+				// Retry with a delay.
+				return gomega.TryAgainAfter(retryAfter)
+			}
+			// Give up, some other error occurred.
+			return gomega.StopTrying(fmt.Sprintf("update %s", resourceType)).Wrap(err)
+		}).Should(gomega.Succeed())
+		verifyObject("update", desiredUpdatedObject, existingObject)
+		patchData := patch.GetData(existingObject)
+
+		ginkgo.By(fmt.Sprintf("Patching status with %s:\n%s", patch.Type, string(patchData)))
+		options := metav1.PatchOptions{FieldValidation: "Strict"}
+		switch patch.Type {
+		case types.ApplyYAMLPatchType, types.ApplyCBORPatchType:
+			options.FieldManager = "test-apply"
+			options.Force = ptr.To(true)
+		}
+		existingObject, err = client.Patch(ctx, desiredInitialObject.GetName(), patch.Type, patchData, options, "status")
+		framework.ExpectNoError(err, "patch %s %s status", patch.Type, resourceType)
+		verifyObject(fmt.Sprintf("patch %s status", patch.Type), desiredUpdatedObjectWithStatus, existingObject)
+	}
+
+	// Use the label as selector in List and DeleteCollection calls.
+	ginkgo.By(fmt.Sprintf("Listing %s collection with label selector %s", gvr, listOptions.LabelSelector))
+	items, err := client.List(ctx, listOptions)
+	framework.ExpectNoError(err, "list %s", resourceType)
+	gomega.Expect(items.Items).Should(gomega.HaveLen(1), "Should have listed exactly the test resource.")
+	verifyObject("list", desiredUpdatedObject, &items.Items[0])
+
+	if tc.IsNamespaced() {
+		ginkgo.By(fmt.Sprintf("Listing %s without namespace and with label selector %s", gvr, listOptions.LabelSelector))
+		items, err := resourceClient.List(ctx, listOptions)
+		framework.ExpectNoError(err, "list %s in all namespaces", resourceType)
+		gomega.Expect(items.Items).Should(gomega.HaveLen(1), "Should have listed exactly the test resource in all namespaces.")
+		verifyObject("list all namespaces", desiredUpdatedObject, &items.Items[0])
+	}
+
+	ginkgo.By(fmt.Sprintf("Deleting %s collection with label selector %s", gvr, listOptions.LabelSelector))
+	err = client.DeleteCollection(ctx, metav1.DeleteOptions{}, listOptions)
+	framework.ExpectNoError(err, "delete collection of %s", resourceType)
+	ensureNotFound(ctx, getResource)
+}
+
+// ensureNotFound ensures that the error is NotFound.
+func ensureNotFound(ctx context.Context, get func(context.Context) (*unstructured.Unstructured, error)) {
+	ginkgo.GinkgoHelper()
+	ginkgo.By("Checking for existence")
+	gomega.Eventually(ctx, func(ctx context.Context) error {
```

Comment:
nit: add a comment explaining this allows objects to be finalized asynchronously before actually being removed